### PR TITLE
AI Camera Jammer Traitor/Thief Item

### PR DESCRIPTION
**Pull request by JustinWinningham**
## About the PR
Added Traitor / Thief item - AI Camera jammer

## Why / Balance
Gives traitors a way to block out AI cameras for a short time, currently traitors have no counterplay against AI eye and its almost impossible to tell where AI is looking at any given moment. Does not block TV feeds like those in security office from viewing cameras normally.

## Technical details
Adds a check to AI view system to check if camera is jammed before granting view via ActiveAiCameraJammerComponent and StationAiVisionSystem.

Uses event driven approach to reduce server load, and throttles check rate to further improve performance. Side effect is we have to track the event start position then find the current offset to know what cameras to jam vs triggering every single update - hybrid approach.

## Media
https://www.youtube.com/watch?v=q-WLVWcQLwQ

## Requirements

- [X] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [X] I have added media to this PR or it does not require an in-game showcase.

## Breaking changes
N/A

**Changelog**

:cl:
- add: Added AI Camera Jammer

### CODE REVIEW

**Comment by Princess-Cheeseballs** [comment]
Old branch. Also this is quite pricey and should probably be merged into the RadioJammer.

**Comment by JustinWinningham** [comment]
> Old branch. Also this is quite pricey and should probably be merged into the RadioJammer.

I really hate dealing with git. I created a new branch from master when I started this but it still shows the old commit history from a long time ago. Not sure how to fix that. I could merge into the radio jammer but this doesnt really have anything to do with radios.

File: Content.Server/Silicons/StationAi/AiCameraJammerSystem.cs [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    private void UpdateJammedCameras(Entity<ActiveAiCameraJammerComponent, AiCameraJammerComponent, TransformComponent> jammer)
+    {
+        var range = GetCurrentRange((jammer.Owner, jammer.Comp2));
+        var jammerPos = _transform.GetMapCoordinates(jammer.Owner, jammer.Comp3);
+
+        // Find all cameras in range
+        var camerasInRange = new HashSet<EntityUid>();
+        foreach (var camera in _lookup.GetEntitiesInRange<StationAiVisionComponent>(jammerPos, range))
```
Comment (Princess-Cheeseballs) [inline Content.Server/Silicons/StationAi/AiCameraJammerSystem.cs]: Doing a lookup like this in an update loop is extremely expensive. It would be much cheaper to have the Jammer use an circular AABB with start and end collide events.

**Review by Princess-Cheeseballs — CHANGES_REQUESTED** [review]

**Comment by beck-thompson** [comment]
super cool but yeah I agree with princess it probably should just be combined with the radio jammer 🤔

The only issue is it would make using the radio jammer harder (if you just wanted the jamming part) because the AI could see something is wrong faster

**Comment by Princess-Cheeseballs** [comment]
> super cool but yeah I agree with princess it probably should just be combined with the radio jammer 🤔
> 
> The only issue is it would make using the radio jammer harder (if you just wanted the jamming part) because the AI could see something is wrong faster

We've floated the idea of an event that shuts off cameras randomly. I think both could be good additions but I'd like to see an item that justifies their addition first.

**Comment by beck-thompson** [comment]
> We've floated the idea of an event that shuts off cameras randomly. I think both could be good additions but I'd like to see an item that justifies their addition first.

good idea!

**Comment by JustinWinningham** [comment]
Guess I'll start combining them + implementing that BB driven approach.

File: Content.Server/Silicons/StationAi/AiCameraJammerSystem.cs [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+        // Handle battery drain and throttled position-based updates for held jammers
+        var query = EntityQueryEnumerator<ActiveAiCameraJammerComponent, AiCameraJammerComponent, TransformComponent>();
+        var curTime = _timing.CurTime;
+
+        while (query.MoveNext(out var uid, out var active, out var jammer, out var xform))
+        {
+            // Try to drain battery
+            if (_powerCell.TryGetBatteryFromSlot(uid, out var battery))
+            {
+                if (!_battery.TryUseCharge(battery.Value.AsNullable(), GetCurrentWattage((uid, jammer)) * frameTime))
```
Comment (slarticodefast) [inline Content.Server/Silicons/StationAi/AiCameraJammerSystem.cs]: Don't use charge every single tick, this is very bad for performance as it dirties the batteries every update. You blindly copy pasted this from `JammerSystem` and removed my code comment there that tells you this.

grafik

Instead make it event based by using `RefreshChargeRate`. This makes handling the charge thresholds more complicated though.
I'm currently working on a PR that fixes this for `JammerSystem`.

Also this is a lot of code duplication from `JammerSystem`, which can be avoided.

**Review by slarticodefast — CHANGES_REQUESTED** [review]

**Comment by Codubreku** [comment]
I don't necessarily mind this being roped in with the Radio Jammer, but I caution against the device's mode being adjusted to have camera jamming along with radio jamming. I can see the need to have have separate modes that you toggle through, like the door remote.
I.e. The Hand Ui would say

> Jamming Target:

and the options would be:
1. Radios
2. Cameras
3. Both

**Comment by TeenSarlacc** [comment]
> The only issue is it would make using the radio jammer harder (if you just wanted the jamming part) because the AI could see something is wrong faster

what if it had a menu similar to the intercoms (with the mic and speaker button being separate) that let you toggle cameras and radios separately? maybe even have it consume more power if you do both at the same time?




(EDIT: didn't scroll enough to see the previous comment when I typed this up, oops)

**Comment by jamilnielsen** [comment]
i was originally not gonna say anything thinking "this is gonna be so easy to abuse as an AI so i dont mind this getting merged"
but i guess i should say it just in case, in ss13 the go-to was to have a tool that made ai-eye visible, the reason jamming ai vision was not used was its so easy to sidestep as AI, vision behaving weirdly? check suit sensors, and inform sec, just to mention one.

**Comment by JustinWinningham** [comment]
@ScarKy0 is your update for jammer refactor done? Waiting on that to proceed with my refactor.

**Comment by ScarKy0** [comment]
>  is your update for jammer refactor done? Waiting on that to proceed with my refactor.

Oh I wasnt working on a jammer refactor

**Comment by JustinWinningham** [comment]
Did not rename "RadioJammer" to generic Jammer to avoid breaking maps

**Comment by JustinWinningham** [comment]
Note, need to retest + second set of eyes on testing edge cases would be nice.

**Comment by Princess-Cheeseballs** [comment]
Raising priority on this since it's a desired buff for the jammer, and this behavior is also needed for future radio jammer behavior.

**Comment by JustinWinningham** [comment]
broke something again. converting back to draft...

**Comment by Princess-Cheeseballs** [comment]
Test fail is real.

**Comment by JustinWinningham** [comment]
HNNNG but it wooooooooooooorks on my machine.... will look into it in the next few days when i get time

File: Content.Server/Silicons/StationAi/AiCameraJammerSystem.cs (hunk 2 above)
Comment (JustinWinningham) [reply Content.Server/Silicons/StationAi/AiCameraJammerSystem.cs]: Refactored.

**Comment by JustinWinningham** [comment]
Praise be I’ve been malding over those test failures for so long I was starting to lose hope.

**Comment by Samuka-C** [comment]
@JustinWinningham time to wake up. I am gonna help you fix this PR. do you want me to solve the merge conflicts or you want do it yourself?

also, while i have you here, please update the PR comment to reflect your current objective with this PR. (I assume things changed going by the comments here)
